### ISSUE TITLE
named parameters not recognized with first class functions

### ISSUE THREAD

**Issue by luk-f-a** [numba/numba#5540]
- [x ] I am using the latest ~~released~~ **master** version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

There is a problem recognizing named parameters. The following fails

```python
from numba import njit, types

@njit(types.int64(types.int64))
def foo(x):
    return x+1

@njit
def loop(foos):
    f = foos[0]
    return f(x=1)

loop((foo,))

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.core.typeinfer.CallConstraint object at 0x7f67f39a1a30>.
mismatch of arguments number: 0 vs 1
[1] During: resolving callee type: UndefinedFunctionType[undefined(undefined)20]
```

while this works

```python
@njit
def loop(foos):
    f = foos[0]
    return f(1)

loop((foo,))
```

**Comment by gmarkall** [comment]
I'm presently on this commit:

```
commit ce2139c7dd93127efb04b35f28f4bebc7f44dfd5 (HEAD -> master, numba/master, numba/HEAD)
Merge: 6e3272769 70944d381
Author: Siu Kwan Lam <michael.lam.sk@gmail.com>
Date:   Thu Apr 9 11:10:15 2020 -0500

    Merge pull request #5538 from stuartarchibald/wip/change_log_049rc2
    
    Update CHANGE_LOG for 0.49.0rc2
```

and the output of the reproducer is:

```
$ python repro.py 
/home/gmarkall/numbadev/numba/numba/core/utils.py:509: NumbaExperimentalFeatureWarning: First-class function type feature is experimental
  category=NumbaExperimentalFeatureWarning)
repro.py:9: NumbaExperimentalFeatureWarning: First-class function type feature is experimental
  f = foos[0]
```

Is this fixed already? Or do you see the same problem with a very recent master, @luk-f-a ?

**Comment by luk-f-a** [comment]
@gmarkall , you're right it does not happen on the latest master! I was on something between rc1 and now, but not today's master. Sorry for the noise!

**Comment by luk-f-a** [comment]
@gmarkall , now I realized what happened. The last commit in master disables first-class functions in many cases (including the reproducer). So the error does not appear, but not for being fixed in the first class function implementation, but for not using first-class functions.
I'm changing the reproducer above, so it fails again on the latest master.

**Comment by stuartarchibald** [comment]
CC @pearu, please could you take a look at this? Thanks.

**Comment by pearu** [comment]
I can reproduce the issue and I am afraid that calling first-class functions with named parameters cannot be supported.

Consider for example
```
@njit('int64(int64)')
def foo1(x):  # never used here
    return x + 1

@njit('int64(int64)')
def foo2(y):
    return y + 1

@njit
def bar(funcs):
    f = funcs[0]
    return f(x = 1)

bar((foo2,))
```
There is no clear way to relate the `y` argument of `foo2` to the named parameter `x`.
Notice that the above example would not make sense in pure Python either.

A possible solution would be to attach parameter names to function signatures, for instance:
```
@njit('int64(int64:x)')
def foo2(y):
    return y + 1
```
but even then the mapping of `x` to `y` is obscure, imho.

In conclusion, first-class functions must be called using positional arguments only.

**Comment by sklam** [comment]
For now, i think we should fix it with better error message.
